### DISCUSSION TITLE
Working on automatic pagination

### DISCUSSION THREAD

**Discussion by jmgirard** [surveydown-dev/surveydown#103] [Q&A] [answered]
### Description

I feel like I'm getting close, but the approach used below merely adds a button every 10 items (all on the same page), rather than actually inserting a page break - presumably because page breaks are being handled by the div fences. Any ideas?

In case you want to test: [pid_items.csv](https://github.com/user-attachments/files/17196445/pid_items.csv)

````qmd
---
echo: false
warning: false
---

```{r}
library(surveydown)

# Read items from file
items <- read.csv("pid_items.csv")

# Set response options
options <- c(
  "Very False or Often False" = "0",
  "Sometimes or Somewhat False" = "1",
  "Sometimes or Somewhat True" = "2",
  "Very True or Often True" = "3"
)
```

::: {#survey .sd-page}

```{r}
# Look up total number of items
ni <- nrow(items)

# Set desired number of items per page
ipp <- 10

# Initialize tagList object
out <- shiny::tagList()

# Loop through pages
for (p in 1:(ceiling(ni / ipp))) {
  # Find items on current page
  pitems <- (ipp * (p - 1) + 1):min(p*ipp, ni)
  # For each item on current page
  for (i in pitems) {
    # Add question to tagList
    out <- shiny::tagList(
      out,
      sd_question(
        id = sprintf("bf_%03d", items[[i, "PID5"]]),
        type = "mc",
        label = items[[i, "Stem"]],
        option = options
      )
    )
  }
  # At the end of each page, add next button
  out <- shiny::tagList(out, sd_next())
}

out
```

:::

::: {#end .sd-page}

## End

```{r}
sd_close("Exit Survey")
```

:::
````

**Comment by jhelvy** [comment]
Yes, the page divs are how pages are defined. So you would need to insert a page start with something like `::: {#page1 .sd-page}` and end with `:::`. I have to believe there is a way to do this though. Ultimately the `sd_question()` and `sd_next()` buttons are just returning HTML code, so perhaps there's a way to define the page breaks as strings then use `#| output: asis` in the code chunk setting? Not sure, would have to play with it for a while to see if it's possible.

**Reply by jmgirard** [reply]
Hmm... I can dynamically create div-pages using `asis` but am struggling to get the output of `sd_question()` and `sd_next()` to work under `asis`. I thought `cat(as.character(sd_question(...)))` might work, but this just shows the code without interpreting it.

**Reply by jhelvy** [reply]
Those functions are returned inside `htmltools::HTML()`. Maybe wrap them inside an `as.character()` call?

**Reply by jmgirard** [reply]
> Those functions are returned inside `htmltools::HTML()`. Maybe wrap them inside an `as.character()` call?

Unfortunately not

**Reply by jhelvy** [reply]
Here's another idea. You can create different components of the survey and store those components as variables. Then in another code chunk, you can run those components, which just turns into html. This example manually creates the page start and end div code as well as the code for one question and a next button. Notice that I have to wrap the page start and end divs inside `htmltools::HTML()`. I don't have to do this for `sd_question()` and `sd_next()` because these are already wrapped internally. Maybe you can work with this idea to programmatically construct all of the components in your survey (include the page divs) and then just have on long code chunk at the bottom that inserts the pieces in the right order?

````qmd
---
format: html
echo: false
warning: false
---

```{r}
library(surveydown)
```

```{r}
#| include: false

page1_start <- htmltools::HTML('<div id="page1" class="sd-page">')
page1_end <- htmltools::HTML('</div>')
q1 <- sd_question(
  type  = "text",
  id    = "penguins_other_basic",
  label = "Please specify the other penguin type:"
)
next1 <- sd_next()
```

```{r}
page1_start
q1
next1
page1_end
```

````

**Reply by jmgirard** [reply]
Thanks, I'll give that a shot.

**Comment by jmgirard** [comment, marked as answer]
Got it working. Needed to keep adding parts to the tagList. Thanks for your help.

````qmd
---
echo: false
warning: false
---

```{r}
library(surveydown)
library(hitop)

# Read items from file
items <- read.csv("pid_items.csv")

# Select version
version <- "PID5BF"

# Filter version items
vitems <- items[!is.na(items[[version]]), ]

options <- c(
  "Very False or Often False" = "0",
  "Sometimes or Somewhat False" = "1",
  "Sometimes or Somewhat True" = "2",
  "Very True or Often True" = "3"
)
```

::: {#welcome .sd-page}

# Welcome to our survey!

You will be provided with a list of things different people might say about themselves. We are interested in how you would describe yourself. There are no "right" or "wrong" answers. We’d like you to take your time and read each statement carefully, selecting the response that best describes you.

```{r}
sd_next()
```

:::

```{r}
# Loop up total number of items
ni <- nrow(vitems)

# Set desired number of items per page
ipp <- 10

# Initialize tag list
out <- shiny::tagList()

# Loop through pages
for (p in 1:(ceiling(ni / ipp))) {
  # Find items on current page
  pitems <- (ipp * (p - 1) + 1):min(p*ipp, ni)
  
  # Add page start div opener to tag list
  out <- shiny::tagList(out, htmltools::HTML(
    paste0('<div id="page', p, '" class="sd-page">')
  ))

  # For each item on current page
  for (i in pitems) {
    # Add question to tagList
    out <- shiny::tagList(out, sd_question(
      id = sprintf("bf_%03d", vitems[[i, version]]),
      type = "mc",
      label = vitems[[i, "Stem"]],
      option = options
    ))
  }
  # At the end of each page, add next button
  out <- shiny::tagList(out, sd_next(), htmltools::HTML('</div>'))
}
out
```

::: {#end .sd-page}

## End

Here are your scores:

```{r}
shiny::tableOutput('my_table')
```


```{r}
sd_close("Exit Survey")
```

:::

````

**Reply by jhelvy** [reply]
Cool! Glad this is now working. I might update an example of this on the documentation.

**Reply by jmgirard** [reply]
If you're open to contributions, I can also make a demo or vignette on generating surveys from a spreadsheet of items with automatic pagination. But I'm also pretty busy so happy to defer to you all to do it.

**Reply by jhelvy** [reply]
Making a simple demo would probably be more helpful than a lengthy vignette. We're pretty slammed too, but if you want to convert what you've got into a simple demo, I can make a new demo repo for this and then you can just send a PR there.
